src/GafferMLModule/GafferMLModule.cpp

```diff
@@ -37,6 +37,125 @@
 
 #include "boost/python.hpp"
 
+#include "GafferML/Tensor.h"
+
+#include "IECorePython/RunTimeTypedBinding.h"
+
+#include "boost/python/suite/indexing/container_utils.hpp"
+
+#include "fmt/format.h"
+
+using namespace boost::python;
+using namespace IECore;
+using namespace GafferML;
+
+namespace
+{
+
+TensorPtr tensorConstructorWrapper( const DataPtr &data, object pythonShape )
+{
+	if( pythonShape != object() )
+	{
+		std::vector<int64_t> shape;
+		boost::python::container_utils::extend_container( shape, pythonShape );
+		return new Tensor( data, shape );
+	}
+	else
+	{
+		return new Tensor( data );
+	}
+}
+
+list tensorShapeWrapper( const Tensor &tensor )
+{
+	list result;
+	for( const auto &x : tensor.shape() )
+	{
+		result.append( x );
+	}
+	return result;
+}
+
+std::string tensorRepr( const Tensor &tensor )
+{
+	if( !tensor.value() )
+	{
+		// The most common use of `repr()` is in serialising the
+		// empty default value for TensorPlug constructors. Make sure
+		// we have a nice clean serialisation for that.
+		return "GafferML.Tensor()";
+	}
+	else
+	{
+		// We don't have a good `repr()` for this - just return a default one
+		// and the ValuePlugSerialiser will attempt a base 64 encoding instead.
+		return fmt::format( "<GafferML._GafferML.Tensor object at {}>", (void *)&tensor );
+	}
+}
+
+template<typename T>
+object tensorGetItemTyped( const Tensor &tensor, const std::vector<int64_t> &location )
+{
+	return object(
+		const_cast<Ort::Value &>( tensor.value() ).At<T>( location )
```

Comment:
No - as documented in `Tensor.h`, we enforce that Tensors are immutable after construction. Here we need the `const_cast()` because the Ort API doesn't provide a non-const flavour of `At()`. But this is invisible from Python - what we return in the end is a copy of the element, not a reference to it.

Reply:
ah I see!